employee-rostering-gwtui/src/test/java/org/optaweb/employeerostering/gwtui/client/pages/shiftroster/ShiftEditFormTest.java

```diff
@@ -0,0 +1,496 @@
+        when(from.getValue()).thenReturn(LATER_LOCAL_DATE_TIME);
+        when(to.getValue()).thenReturn(EARILER_LOCAL_DATE_TIME);
+        when(from.reportValidity()).thenReturn(true);
+        when(to.reportValidity()).thenReturn(true);
+
+        boolean out = testedShiftEditForm.updateShiftFromWidgetsIfValid(newShiftView);
+
+        // Assert any changes are rolled back
+        assertThat(newShiftView).isEqualTo(oldShiftView);
+        assertThat(out).isFalse();
+        verify(from).reportValidity();
+        verify(to).reportValidity();
+        verify(notificationFactory).showError(any());
+    }
+
+    @Test
+    public void testInvalidWidgetUpdateShiftFromWidgetsIfValid() {
+        ShiftView oldShiftView = shiftView;
+
+        // Clone the shift view since this one will may be modified
+        ShiftView newShiftView = new ShiftView(TENANT_ID, spotList.get(SELECTED_SPOT),
+                                               EARILER_LOCAL_DATE_TIME,
+                                               LATER_LOCAL_DATE_TIME);
+
+        final int NEW_SELECTED_SPOT = 3;
+        final int NEW_SELECTED_EMPLOYEE = 1;
+
+        pinned.checked = true;
+        when(spotSelect.getSelectedValue()).thenReturn(spotList.get(NEW_SELECTED_SPOT).getId().toString());
+        when(employeeSelect.getSelectedValue()).thenReturn(employeeList.get(NEW_SELECTED_EMPLOYEE).getId().toString());
+        when(from.getValue()).thenReturn(null);
+        when(to.getValue()).thenReturn(EARILER_LOCAL_DATE_TIME);
+        when(from.reportValidity()).thenReturn(false);
+        when(to.reportValidity()).thenReturn(true);
+
+        boolean out = testedShiftEditForm.updateShiftFromWidgetsIfValid(newShiftView);
+
+        // Assert any changes are rolled back
+        assertThat(newShiftView).isEqualTo(oldShiftView);
+        assertThat(out).isFalse();
+        verify(notificationFactory).showError(any());
+    }
+
+    @Test
+    public void testValidUpdateShiftFromWidgetsIfValid() {
+        final int NEW_SELECTED_SPOT = 3;
+        final int NEW_SELECTED_EMPLOYEE = 1;
+
+        pinned.checked = true;
+        when(spotSelect.getSelectedValue()).thenReturn(spotList.get(NEW_SELECTED_SPOT).getId().toString());
+        when(employeeSelect.getSelectedValue()).thenReturn(employeeList.get(NEW_SELECTED_EMPLOYEE).getId().toString());
+        when(from.getValue()).thenReturn(EARILER_LOCAL_DATE_TIME.plusDays(2));
+        when(to.getValue()).thenReturn(LATER_LOCAL_DATE_TIME.plusDays(2));
+        when(from.reportValidity()).thenReturn(true);
+        when(to.reportValidity()).thenReturn(true);
+
+        boolean out = testedShiftEditForm.updateShiftFromWidgetsIfValid(shiftView);
+
+        verify(spotSelect).getSelectedValue();
+        verify(employeeSelect).getSelectedValue();
+
+        verify(from).reportValidity();
+        verify(to).reportValidity();
+        verify(from).getValue();
+        verify(to).getValue();
```

Comment:
All of these `verify()` statements are redundant as the calls to all of the method above are implicitly covered by the assertions below. Not a mistake, just a FYI and a suggestion how to make the test more concise.